### DISCUSSION TITLE
Cannot run renovate on a gomod project with private dependencies

### DISCUSSION THREAD

**Discussion by xtreme-shane-lattanzio** [renovatebot/renovate#17062] [Request Help]
I have been trying to get renovate running in the docker image on a project with private go mod dependencies. As per this page: https://docs.renovatebot.com/getting-started/private-packages/#gomod, I have set the GIT_CONFIG_KEY_0, GIT_CONFIG_VALUE_0 and GIT_CONFIG_COUNT variables as follows:

GIT_CONFIG_KEY_0=url.https://user:password@gitlab.com/.insteadOf
GIT_CONFIG_VALUE_0=https://gitlab.com/
GIT_CONFIG_COUNT=1

After doing this, when I run renovate I get many errors for `error: missing config key GIT_CONFIG_KEY_0\n\tfatal: unable to parse command-line config\n"` Am I doing something wrong? If I manually set the gitconfig, everything seems to work. Any help would be appreciated. Thanks!

**Comment by Gabriel-Ladzaretti** [comment]
@nabeelsaabna can you assist on this?
might be related to https://github.com/renovatebot/renovate/issues/17045 ?

**Comment by nabeelsaabna** [comment]
Hi @xtreme-shane-lattanzio

the `GIT_CONFIG_KEY_0` and `GIT_CONFIG_VALUE_0` AND `GIT_CONFIG_COUNT` are added automatically by renovate from `hostRules`.

Your `renovate.json` (or `config.js`) should include something like below `hostRules` and `customEnvVariables`:
```
{
	"$schema": "https://docs.renovatebot.com/renovate-schema.json",
	"hostRules": [
		{
			"matchHost": "https://gitlab.com/",
			"hostType": "go",
			"token": "your security token"
		}
	],
	"customEnvVariables": {
		"GOPRIVATE": "gitlab.com/nabeelsaabna",
		"GONOSUMDB": "gitlab.com/nabeelsaabna",
		"GONOPROXY": "gitlab.com/nabeelsaabna"
	}
}
```

**Reply by xtreme-shane-lattanzio** [reply]
Thanks for the reply. Does that mean they cannot be manually added? I haven't modified the renovate json yet and was trying to avoid it but if its best I can. Is the token a personal access token, deploy key, something else?

**Reply by nabeelsaabna** [reply]
Don't believe renovate uses (or should use) the environment variables to do the "look-up" (which is http based), that's why it requires the host-rules.
For the `customEnvVariables` it can configured in the environment instead. 

Again, you can edit the `config.js` for all repositories instead of editing/exposing tokens in each repository under `renovate.json`

**Reply by xtreme-shane-lattanzio** [reply]
Okay thanks. I am struggling with this though. I think I'm missing something basic. I have committing the following renovate json files
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "hostRules": [
    {
      "matchHost": "https://gitlab.com/",
      "hostType": "go",
      "username": "me",
      "password": "<personal_access_token>"
    }
  ]
}
```
OR
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "hostRules": [
    {
      "matchHost": "https://gitlab.com/",
      "hostType": "go",
      "token": "<personal_access_token>"
    }
  ]
}
```

and I still get access issues. I have also tried committing the config.js to the repo or just creating this and setting the RENOVATE_CONFIG_FILE env var:
```
module.exports = {
    hostRules: [
        {
            matchHost: 'https://gitlab.com/',
            hostType: 'go',
            username: 'me',
            password: '<personal_access_token>'
        }
    ]
};
```

This also does not work. If it did I would also need a way to hide the access token so I could commit these files and I am not sure how to do that either. Any pointers on this part? Am I close? Thanks again!!!

**Reply by xtreme-shane-lattanzio** [reply]
When I use my personal access token as the RENOVATE_TOKEN, the log seems to be fine even without a config file although it will not longer rebase the existing PRs. However, if I use the RENOVATE_TOKEN as the access token to the repo, I get issues. I am just trying to understand why this does not work.

I think I found another way. If I create a config.js, the hostType may need to be gitlab. I also needed to change the host to the full URL of the dependency though because renovate was having 409 errors when trying with just the base gitlab URL. Does that sound right? However, when it tries to run the `go get` command it still returns on stderr but it manages to update the go.sum. Any idea why it logs to stderr? Am I supposed to change it to gitlab instead?  If I don't, I get auth issues when the `go get` runs. (fatal: Authentication failed for.....) 

But then I realized there are actually two issues. The first was with the artifact updating which changing the hostType seemed to fix. The second is getting a 404 on the Gitlab API as well as the go Datasource for that dependency when it tried to look up the dependency early on in the log. Is this a problem?  I find it strange that for the single package, it is looking up three different things for it:

```
DEBUG: Go lookup source url (repository=org/mainrepo)
       "goModule": "gitlab.com/org/deprepo",
       "goSourceUrl": "https://gitlab.com/org/deprepo"
DEBUG: GitLab API 404 (repository=org/mainrepo)
       "url": "https://gitlab.com/api/v4/projects/org%2Fdeprepo/repository/tags?per_page=100"
DEBUG: Datasource 404 (repository=org/mainrepo)
       "datasource": "go",
       "packageName": "gitlab.com/org/deprepo,
       "url": "https://gitlab.com/api/v4/projects/org%2Fdeprepo/repository/tags?per_page=100"
```
Again this happens when I use the top repo access token as the RENOVATE_TOKEN and then try to overwrite the token for the dependency with something like:
```
module.exports = {
  hostRules: [
    {
      matchHost: 'https://gitlab.com/org/deprepo',
      hostType: 'gitlab',
      token: 'PERSONAL_ACCESS_TOKEN'
    }
  ]
};
```

**Reply by xtreme-shane-lattanzio** [reply]
@nabeelsaabna I actually wonder if its easier just to manually have a git config command before running renovate but that still gives the 404 errors while writing out to stderr

**Reply by nabeelsaabna** [reply]
Hi @xtreme-shane-lattanzio 
I'm a bit overwhelmed by the info!
it might be helpful read the discussion https://github.com/renovatebot/renovate/discussions/16906 and the issue that resulted from it https://github.com/renovatebot/renovate/issues/17045 to better understand host-rules

**Comment by xtreme-shane-lattanzio** [comment]
Hi @nabeelsaabna ! I know I dumped a lot in there lol Those links are more about how the things convert to git config commands especially when there are multiple env vars. Really I am down to a few questions:

1. I use a repository access token as the RENOVATE_TOKEN. Can I use a personal access token for the dependency repo? Right now it looks like I can only if I set the hostType to gitlab and not go but I still get the 404 errors as listed above. I get this same issue if I manually set the git config with my username and the personal access token.
2. Should the hostType even be gitlab? All of the examples I see have the hostType as go but I still get auth issues when doing that
3. Username and password will never work right? It needs to be a token?
4. When running renovate, if I get around the auth errors, the `go get` still seems to write to stderr but continues. Is there a way to find out why? The 404 issues will also show in this case

A lot of this is me needing to simply understand how this works or doesn't work. Renovate seems to go green even if there is an issue so I am combing through the logs to find red flags. I guess I am wondering for best practices in a case that I have a private repo with private gitlab dependencies in the go mod.

**Reply by nabeelsaabna** [reply]
Learning about `hostRules` and loving the golang programming language makes me curious enough to try it myself !!
will update with the findings

**Comment by nabeelsaabna** [comment]
I'm running on a github.com repository using below `config.js`

```
module.exports = {
	token: 'ghp_github_token',
	platform: 'github',
	gitAuthor: "Renovate Bot <***@github.com>",
	dependencyDashboard: true,
	repositories: [
		'nabeelys/go-public-for-gitlab-private'
	],	
	hostRules: [
		{
			matchHost: "https://gitlab.com",
			token: "glpat-gitlab_token",
			username: "nabeelsaabna"
		}
	]
};
```

the lookup is working fine without any issues and it's managing to find the `"newVersion": "v1.9.0"`:

<details>

<summary>logs</summary>

```
DEBUG: packageFiles with updates (repository=nabeelys/go-public-for-gitlab-private)
       "config": {
         "gomod": [
           {
             "packageFile": "go.mod",
             "constraints": {"go": "^1.17"},
             "deps": [
               {
                 "managerData": {"lineNumber": 2},
                 "depName": "go",
                 "depType": "golang",
                 "currentValue": "1.17",
                 "datasource": "golang-version",
                 "versioning": "npm",
                 "rangeStrategy": "replace",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "1.19.0",
                     "newValue": "1.19",
                     "releaseTimestamp": "2022-08-02T00:00:00.000Z",
                     "newMajor": 1,
                     "newMinor": 19,
                     "updateType": "minor",
                     "isRange": true,
                     "branchName": "renovate/go-1.x"
                   }
                 ],
                 "warnings": [],
                 "sourceUrl": "https://github.com/golang/go",
                 "homepage": "https://go.dev/",
                 "currentVersion": "1.17.13",
                 "isSingleVersion": false
               },
               {
                 "managerData": {"lineNumber": 5, "multiLine": true},
                 "depName": "gitlab.com/nabeelsaabna/go-private-nabeel",
                 "depType": "require",
                 "currentValue": "v1.8.0",
                 "datasource": "go",
                 "depIndex": 1,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "v1.9.0",
                     "newValue": "v1.9.0",
                     "releaseTimestamp": "2022-08-11T08:14:42.000Z",
                     "newMajor": 1,
                     "newMinor": 9,
                     "updateType": "minor",
                     "branchName": "renovate/gitlab.com-nabeelsaabna-go-private-nabeel-1.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "semver",
                 "sourceUrl": "https://gitlab.com/nabeelsaabna/go-private-nabeel",
                 "currentVersion": "v1.8.0",
                 "isSingleVersion": true,
                 "fixedVersion": "v1.8.0"
               },
               {
                 "managerData": {"lineNumber": 6, "multiLine": true},
                 "depName": "github.com/google/go-cmp",
                 "depType": "require",
                 "currentValue": "v0.5.4",
                 "datasource": "go",
                 "depIndex": 2,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "v0.5.8",
                     "newValue": "v0.5.8",
                     "releaseTimestamp": "2022-04-26T20:49:16.000Z",
                     "newMajor": 0,
                     "newMinor": 5,
                     "updateType": "patch",
                     "branchName": "renovate/github.com-google-go-cmp-0.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "semver",
                 "sourceUrl": "https://github.com/google/go-cmp",
                 "currentVersion": "v0.5.4",
                 "isSingleVersion": true,
                 "fixedVersion": "v0.5.4"
               }
             ]
           }
         ]
       }
```

</details>

then command is failing with error:
```
"cmd": "/bin/sh -c go get -d -t ./...",
         "stderr": "go: gitlab.com/nabeelsaabna/go-private-nabeel@v1.9.0: reading gitlab.com/nabeelsaabna/go-private-nabeel/go.mod at revision v1.9.0: git ls-remote -q origin in /go/pkg/mod/cache/vcs/99104e893ebd5f8fa6a7e86253de573168358f72eb97491a5807a1ea4d9d692b: exit status 128:\n\tfatal: could not read Password for 'https://**redacted**@gitlab.com': terminal prompts disabled\n",
         "stdout": ""
```

after a bit of investigation it turnout the git config:
```
             "GIT_CONFIG_KEY_3": "url.https://**redacted**@gitlab.com/.insteadOf",
             "GIT_CONFIG_VALUE_3": "ssh://**redacted**@gitlab.com/",
             "GIT_CONFIG_KEY_4": "url.https://**redacted**@gitlab.com/.insteadOf",
             "GIT_CONFIG_VALUE_4": "git@gitlab.com:",
             "GIT_CONFIG_KEY_5": "url.https://**redacted**@gitlab.com/.insteadOf",
             "GIT_CONFIG_VALUE_5": "https://gitlab.com/"
``` 

is using the github style: 
`url.https://glpat-gitlab_token@gitlab.com/.insteadOf`

while it should be with username: 
`url.https://nabeelsaabna:glpat-gitlab_token@gitlab.com/.insteadOf`

need to open an issue to fix the gitlab GIT_CONFIG to include the username whenever provided

**Reply by nabeelsaabna** [reply]
@Gabriel-Ladzaretti can you please open an issue for this ?

**Reply by Gabriel-Ladzaretti** [reply]
sure

**Reply by Gabriel-Ladzaretti** [reply]
- https://github.com/renovatebot/renovate/issues/17137

**Reply by xtreme-shane-lattanzio** [reply]
@nabeelsaabna Interesting! Thanks for looking into this. The only difference in my testing is I was using username and password (not username and token) and putting the token in the password field. I was getting the password issue you put. However, when I used just token (with a personal access token), I would get simply authentication failed errors.

I also notice you put in no hostType in the hostRules. I never tried that but as I said before, I could only seem to get farther when using the type as gitlab and not go. Either way I would still write to stderr when running `go get -d -t ./...` although the error was not clear and just looked like the output of the command (not an exit error like you put). When this happened though, there were those 404 errors I was talking about.

Do you think this username/passsword combo would fix these things? For now my workaround was to put my personal access token (that also has access to the dependency) in for the RENOVATE_TOKEN. What I wanted to do was to have a repository access token there, and then create a config.js with a different repository access token for the dependency. Both of these repos are gitlab.

Thanks again for following up!

**Reply by xtreme-shane-lattanzio** [reply]
Hey @nabeelsaabna ! Im not sure I followed the PR completely. Given my setup of using a repo token as a `RENOVATE_TOKEN` and wanting to get a private gomodule working, what should my config.js be? Would it be 
```
module.exports = {
    hostRules: [
        {
            matchHost: 'https://gitlab.com/org/gomodulerepo',
            hostType: 'gitlab',
            token: <go module access token>
        }
    ]
};
```

Or something else? Before you said the hostType should be go. Please let me know and thanks for working on this!!

**Reply by nabeelsaabna** [reply]
Hi @xtreme-shane-lattanzio 

As I understand so far you have:

-  repo to run on: `https://gitlab.com/org-a/go-module-a` accessed by the default renovate token
- a different module `go-module-b` accessed by a new token from host-rules, is it under the same `org-a` ?

What is the result of hostRule config provide in the latest example 
1. failure during lookup ?
2. failure during artifact update ?

**Reply by xtreme-shane-lattanzio** [reply]
Hey @nabeelsaabna! Actually both repos are under different orgs but I can get access to both but yes the repo to run on is the renovate token and the config is trying to override a different module in a different org.

Before the PR that was merged, if I create a config.js with the hostType gitlab as above, I also needed to change the host to the full URL of the dependency instead of the short URL because renovate was having 409 errors when trying with just the base gitlab URL. When it tries to run the go get command it still returns on stderr but it manages to update the go.sum. Any idea why it logs to stderr? Is the hosttype of gitlab correct here or should it be go? If I put it as go, I got auth issues when the go get runs. (fatal: Authentication failed for.....). So I am just wondering if this is now resolved in this way.

However, there was a second issue where I was getting a 404 on the Gitlab API as well as the go Datasource for that dependency when it tried to look up the dependency early on in the log:

```
DEBUG: Go lookup source url (repository=org/mainrepo)
       "goModule": "gitlab.com/org/deprepo",
       "goSourceUrl": "https://gitlab.com/org/deprepo"
DEBUG: GitLab API 404 (repository=org/mainrepo)
       "url": "https://gitlab.com/api/v4/projects/org%2Fdeprepo/repository/tags?per_page=100"
DEBUG: Datasource 404 (repository=org/mainrepo)
       "datasource": "go",
       "packageName": "gitlab.com/org/deprepo,
       "url": "https://gitlab.com/api/v4/projects/org%2Fdeprepo/repository/tags?per_page=100"
```

At this point I'm just trying to know the recommended way for this with the new PR and I'll try it out.

**Reply by nabeelsaabna** [reply]
when running with `platform: 'gitlab'` a default host-rule is added and will be used for all `gitlab` calls:

```
{
  "matchHost": "gitlab.com",
  "token": "renovate token",
  "hostType": "gitlab"
}
```

to override the authentication to the second organization you need the 2 rules (one for each URL prefix), due to the way below function works (checks for host, host-type and `prioritizeLongestMatchHost`)
https://github.com/renovatebot/renovate/blob/a657c3476e7a6c78034147ae04c71ae81b88bd17/lib/util/host-rules.ts#L122

```
{
	"hostRules": [
		{
			"matchHost": "https://gitlab.com/second-org/",
			"hostType": "gitlab",
			"token": "<second-org access token>"
		},
		{
			"matchHost": "https://gitlab.com/api/v4/projects/second-org/",
			"hostType": "gitlab",
			"token": "<second-org access token>"
		}
	]
}
```

**Reply by xtreme-shane-lattanzio** [reply]
@nabeelsaabna So I tried this in a config.js and tried to run it and it did not work. I still got the HTTP basic auth denied errors. When I changed the token to my personal access token, it seemed to clear up the error but I still see the following:

```
DEBUG: go get command included (repository=main-org/main-repo, branch=renovate/k8s.io-api-0.x)
       "cmd": "go",
       "args": "get -d -t ./..."
DEBUG: Executing command (repository=main-org/main-repo, branch=renovate/k8s.io-api-0.x)
       "command": "go get -d -t ./..."
DEBUG: exec completed (repository=main-org/main-repo, branch=renovate/k8s.io-api-0.x)
       "cmd": "go get -d -t ./...",
       "durationMs": 12912,
       "stdout": "",
       "stderr": "go: downloading github.com/onsi/ginkgo/v2 v2.1.4\ngo: downloading github.com/onsi/gomega v1.20.0\ngo: downloading k8s.io/api v0.24.4\ngo: downloading k8s.io/apimachinery v0.24.4\ngo: downloading k8s.io/client-go v0.24.3\ngo: downloading github.com/go-logr/logr v1.2.3\n**go: downloading gitlab.com/second-org/second-repo v1.2.0\n**go: downloading github.com/go-logr/zapr v1.2.3\ngo: downloading go.uber.org/zap v1.22.0\ngo: downloading github.com/pkg/errors v0.9.1\ngo: downloading github.com/google/go-cmp v0.5.8\ngo: downloading k8s.io/klog/v2 v2.60.1\ngo: downloading k8s.io/utils v0.0.0-20220210201930-3a6ce19ff2f9\ngo: downloading github.com/gogo/protobuf v1.3.2\ngo: downloading github.com/google/gofuzz v1.1.0\ngo: downloading golang.org/x/oauth2 v0.0.0-20211104180415-d3ed0bb246c8\ngo: downloading golang.org/x/net v0.0.0-20220425223048-2871e0cb64e4\ngo: downloading github.com/imdario/mergo v0.3.5\ngo: downloading github.com/spf13/pflag v1.0.5\ngo: downloading golang.org/x/term v0.0.0-20210927222741-03fcf44c2211\ngo: downloading go.uber.org/atomic v1.7.0\ngo: downloading go.uber.org/multierr v1.7.0\ngo: downloading github.com/hashicorp/go-cleanhttp v0.5.1\ngo: downloading github.com/google/uuid v1.1.2\ngo: downloading gopkg.in/yaml.v2 v2.4.0\ngo: downloading github.com/hashicorp/go-retryablehttp v0.7.0\ngo: downloading gopkg.in/yaml.v3 v3.0.1\ngo: downloading gopkg.in/inf.v0 v0.9.1\ngo: downloading sigs.k8s.io/structured-merge-diff/v4 v4.2.1\ngo: downloading golang.org/x/sys v0.0.0-20220422013727-9388b58f7150\ngo: downloading github.com/golang/protobuf v1.5.2\ngo: downloading github.com/google/gnostic v0.5.7-v3refs\ngo: downloading golang.org/x/time v0.0.0-20220210224613-90d013bbcef8\ngo: downloading sigs.k8s.io/yaml v1.2.0\ngo: downloading github.com/davecgh/go-spew v1.1.1\ngo: downloading cloud.google.com/go v0.81.0\ngo: downloading google.golang.org/appengine v1.6.7\ngo: downloading sigs.k8s.io/json v0.0.0-20211208200746-9f7c6b3444d2\ngo: downloading google.golang.org/protobuf v1.28.0\ngo: downloading k8s.io/kube-openapi v0.0.0-20220328201542-3ee0da9b0b42\ngo: downloading golang.org/x/text v0.3.7\ngo: downloading github.com/json-iterator/go v1.1.12\ngo: downloading github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822\ngo: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd\ngo: downloading github.com/modern-go/reflect2 v1.0.2\ngo: downloading github.com/emicklei/go-restful v2.9.5+incompatible\ngo: downloading github.com/go-openapi/swag v0.19.14\ngo: downloading github.com/go-openapi/jsonreference v0.19.5\ngo: downloading github.com/go-openapi/jsonpointer v0.19.5\ngo: downloading github.com/PuerkitoBio/purell v1.1.1\ngo: downloading github.com/mailru/easyjson v0.7.6\ngo: downloading github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578\ngo: downloading github.com/josharian/intern v1.0.0\ngo: upgraded k8s.io/apimachinery v0.24.3 => v0.24.4\n"
```

Notice it is writing something to stderr. Any idea why this might be? If I run the go get command myself locally after checking out the renovate branch, there is not an issue. There are also no warnings on the PR itself so you can only see this issue in the log :S

**Reply by nabeelsaabna** [reply]
the `go` command always writes messages like `go: downloading <module>` to `stderr` (used for debug messages, not only errors)
the `stderr` is actually a good thing in this example

**Reply by xtreme-shane-lattanzio** [reply]
@nabeelsaabna Ah I see! Thanks for the clarification. I also think I got the repo token in the config.js working. My last one must have been bad. 

Last question! I am also wondering if there is a way to do this with SSH keys or does it have to be done with access tokens?

**Reply by nabeelsaabna** [reply]
for that maybe better to open a new discussion with suitable title for someone with knowledge to help!
